### PR TITLE
Texturing

### PR DESCRIPTION
**Pull request by timoffex**
Significant changes to rendering pipeline: most rendering code moved outside of MeshView into subtypes of AbstractRenderer. Currently, only the SimpleTexturedRenderer is provided.

This PR also includes the 1x1 map fix.

### CODE REVIEW

File: WallsAndHoles/WallsAndHoles.pro.orig
Comment (Bdtrotte) [inline WallsAndHoles/WallsAndHoles.pro.orig]: Stray file!

File: WallsAndHoles/abstractmeshviewcamera.h
Comment (Bdtrotte) [inline WallsAndHoles/abstractmeshviewcamera.h:16]: Any good reason to not just have AbstractTool inherit QObject?

File: WallsAndHoles/abstractrenderer.h
Comment (Bdtrotte) [inline WallsAndHoles/abstractrenderer.h]: doing
```
while ((GLenum error = glGetError()) != GL_NO_ERROR) { 
    switch (error) {
         ... 
    }
 }
```
could be cleaner

File: WallsAndHoles/simpletexturedrenderer.cpp
Comment (Bdtrotte) [inline WallsAndHoles/simpletexturedrenderer.cpp]: Encapsulation! Maybe store the texture above the if (!mImages...) so you don't have to access the map again (And make it cleaner)

File: WallsAndHoles/simpletexturedrenderer.cpp
Comment (Bdtrotte) [inline WallsAndHoles/simpletexturedrenderer.cpp:105]: glEnable(GL_CULL_FACE);

File: WallsAndHoles/simpletexturedrenderer.cpp
Comment (Bdtrotte) [inline WallsAndHoles/simpletexturedrenderer.cpp:125]: why not `Texture t : mTexturesToObjects.keys()` ?

File: WallsAndHoles/simpletexturedrenderer.h
Comment (Bdtrotte) [inline WallsAndHoles/simpletexturedrenderer.h]: I don't think this function should exist. Shouldn't it be updated when an object is removed?

File: WallsAndHoles/editor.cpp.orig
Comment (Bdtrotte) [inline WallsAndHoles/editor.cpp.orig]: Stray file

File: WallsAndHoles/editor.cpp
Comment (Bdtrotte) [inline WallsAndHoles/editor.cpp]: I think delete on a nullptr is safe

File: WallsAndHoles/map2mesh.cpp.orig
Comment (Bdtrotte) [inline WallsAndHoles/map2mesh.cpp.orig]: STRAY FILE (Tima is a stat buffer)

File: WallsAndHoles/map2mesh.h.orig
Comment (Bdtrotte) [inline WallsAndHoles/map2mesh.h.orig]: rm *.orig

File: WallsAndHoles/meshview.cpp
Comment (Bdtrotte) [inline WallsAndHoles/meshview.cpp:189]: Just remove these

File: WallsAndHoles/objtools.h
Comment (Bdtrotte) [inline WallsAndHoles/objtools.h:1]: an implementation file would be nice. And probably wrap it in a namespace

**Review by Bdtrotte — CHANGES_REQUESTED** [review]
Nice stuff.

A few rather trivial changes requested, and some stray files that need be removed.

Get it merged with master.

I still want to do some more testing in the actual app before merge (I was getting some strange behavior). and look through some of the code a bit more. Particularly, I'm pretty sure that when a tile changes, its mesh is just removed, and a new one is added, which is of course fine, but maybe there could be a special route if just the texture/material changes so all the data doesn't have to be changed, but just the uniforms it is rendered with. Maybe those changes could come in in the next commit that adds textures and materials to TileTemplates.

In any case, LOTS of work which is greatly appreciated. We should be able to merge by the end of today

File: WallsAndHoles/abstractmeshviewcamera.h
Comment (timoffex) [reply WallsAndHoles/abstractmeshviewcamera.h:16]: Can't think of any signals that an AbstractTool would have. Maybe deactivate() and activate() could be slots, but they are probably only ever going to be called directly from a ToolManager. And (I think) mouse events in QWidgets don't have corresponding signals, so making those functions into slots wouldn't make any code simpler.

File: WallsAndHoles/abstractrenderer.h
Comment (timoffex) [reply WallsAndHoles/abstractrenderer.h]: Will change, although GLenum error still has to be declared outside of the while loop.

File: WallsAndHoles/simpletexturedrenderer.cpp
Comment (timoffex) [reply WallsAndHoles/simpletexturedrenderer.cpp]: I can make it cleaner, but I am not sure what you mean by encapsulation here.

File: WallsAndHoles/simpletexturedrenderer.cpp
Comment (timoffex) [reply WallsAndHoles/simpletexturedrenderer.cpp:125]: https://www.dvratil.cz/2015/06/qt-containers-and-c11-range-based-loops/
because it might be inefficient. Then again, according to https://www.kdab.com/goodbye-q_foreach/ it may be getting deprecated.

File: WallsAndHoles/simpletexturedrenderer.h
Comment (timoffex) [reply WallsAndHoles/simpletexturedrenderer.h]: Yes, it should be updated. I wrote this function just in case, before I had a signals/slots system that worked. I will remove it.

File: WallsAndHoles/editor.cpp.orig
Comment (timoffex) [reply WallsAndHoles/editor.cpp.orig]: They get created when I use git mergetool, and GitHub Desktop automatically adds them. Sometimes I forget to remove them.

File: WallsAndHoles/meshview.cpp
Comment (timoffex) [reply WallsAndHoles/meshview.cpp:189]: I can remove load(), but I can't remove save() yet because it is used by Editor::exportMapMesh().

Comment (timoffex) [reply WallsAndHoles/meshview.cpp:189]: I'll hold off on doing that though until I get time to work on a nice abstract save/load system.

File: WallsAndHoles/abstractmeshviewcamera.h
Comment (Bdtrotte) [reply WallsAndHoles/abstractmeshviewcamera.h:16]: But, it would allow other tools to use signals and slots, and as far as Im aware, there isn't much overhead for making it inherit QObject

File: WallsAndHoles/simpletexturedrenderer.cpp
Comment (Bdtrotte) [reply WallsAndHoles/simpletexturedrenderer.cpp]: I think it's one of the buzz words from class, maybe I'm miss using it...

**Pull request merged**
